lms/static/scripts/frontend_apps/utils/api.ts

```diff
@@ -191,6 +192,100 @@ export function urlPath(strings: TemplateStringsArray, ...params: string[]) {
 export function useAPIFetch<T = unknown>(
   path: string | null,
   params?: QueryParams,
+): FetchResult<T> {
+  // We generate a URL-like key from the path and params, but we could generate
+  // something simpler, as long as it encodes the same information. The auth
+  // token is not included in the key, as we assume currently that it does not
+  // change the result.
+  const paramStr = recordToQueryString(params ?? {});
+  return useAPIFetchWithKey(path ? `${path}${paramStr}` : null, path, params);
+}
+
+export type PolledAPIFetchOptions<T> = {
+  /** Path for API call */
+  path: string;
+  /** Query params for API call */
+  params?: QueryParams;
+  /** Determines if, based on the result, the request should be retried */
+  shouldRetry: (result: FetchResult<T>) => boolean;
+
+  /**
+   * Amount of ms after which a retry should happen, if shouldRetry() returns
+   * true.
+   * Defaults to 500ms.
+   */
+  retryAfter?: number;
+
+  /** Test seam */
+  _setTimeout?: typeof setTimeout;
+  /** Test seam */
+  _clearTimeout?: typeof clearTimeout;
+};
+
+/**
+ * Hook that fetches data using authenticated API requests, allowing it to be
+ * retried periodically.
+ *
+ * @return The FetchResult with an extra `retries` property that represents the
+ *         amount of times the call was retried after the first one.
+ */
+export function usePolledAPIFetch<T = unknown>({
+  path,
+  params,
+  shouldRetry: unstableShouldRetry,
+  retryAfter = 500,
```

Comment:
For the current use case we expect that grade sync will finish after some not-too-long period of time (a few seconds, maybe a few minutes in the worst case). I suggest we set a limit on the maximum number of refreshes.

Reply:
> I suggest we set a limit on the maximum number of refreshes.

I think I'm going to leave this for later, because it's hard to know what would be a sensible default here.

When you are retrying an error, it's reasonable to retry a few times and that's it. In here, if there's no error, you should retry until whatever you are waiting for has finished, and that could take seconds or minutes, so the amount of retries is directly related with how often are you refreshing.

Also, if you stop refreshing, should that be reflected in the UI somehow, so that it doesn't fell as if it is "malfunctioning"? A message suggesting to reload the page manually?

Perhaps another option would be to reduce the frequency if the refresh takes more time. So first 20 attempts are done every half a second, then every second, then every 2 seconds, and so on.

So since this is not super clear, and consumers can still handle this (see the example in the PR description, where only 5 refreshes happen), I'm going to leave it with indefinite refreshes for now.

Reply:
> Perhaps another option would be to reduce the frequency if the refresh takes more time. So first 20 attempts are done every half a second, then every second, then every 2 seconds, and so on.

This is a good idea, up to some reasonable limit. Say a maximum of 30 seconds between checks?